python/ray/serve/_private/application_state.py

```diff
@@ -85,35 +109,71 @@ def deploy_obj_ref(self) -> Optional[ObjectRef]:
 
     @property
     def deployments(self) -> List[str]:
-        """Return all deployments name from the application"""
-        if self._deployment_params is None:
+        """Names of all deployments in application."""
+        if self._target_state.deployment_infos is None:
             return []
-        return [params["name"] for params in self._deployment_params]
+        return list(self._target_state.deployment_infos.keys())
+
+    def recover_target_state_from_checkpoint(self, checkpoint_data):
+        logger.info(f"Recovering target state for app {self._name} from checkpoint.")
+        self._target_state = checkpoint_data
+
+    def _set_target_state(
+        self,
+        deployment_infos: Optional[Dict[str, DeploymentInfo]] = None,
+        deleting: bool = False,
+    ):
+        if deleting:
+            target_state = ApplicationTargetState(dict(), True)
+            self._status = ApplicationStatus.DELETING
+        else:
+            target_state = ApplicationTargetState(deployment_infos, False)
+            self._status = ApplicationStatus.DEPLOYING
+
+        # Checkpoint ahead
+        self._save_checkpoint_func(writeahead_checkpoints={self._name: target_state})
+        # Set target state
+        self._target_state = target_state
+
+    def _delete_deployment(self, name):
+        self._endpoint_state.delete_endpoint(name)
+        self._deployment_state_manager.delete_deployment(name)
 
     def delete(self):
         """Delete the application"""
-        self._status = ApplicationStatus.DELETING
+        logger.info(f"Deleting application '{self._name}'")
+        self._set_target_state(deleting=True)
+
+    def apply_deployment(
+        self, deployment_name: str, deployment_info: DeploymentInfo
+    ) -> bool:
+        """
+        Deploys a deployment in the application.
 
-    def deploy(self, deployment_params: List[Dict]) -> List[str]:
-        """Deploy the application.
+        Returns: Whether the deployment is being updated.
+        """
+        updating = self._deployment_state_manager.deploy(
+            deployment_name, deployment_info
+        )
+
+        if deployment_info.route_prefix is not None:
+            assert deployment_info.route_prefix.startswith("/")
+            self._endpoint_state.update_endpoint(
+                deployment_name,
+                EndpointInfo(route=deployment_info.route_prefix, app_name=self._name),
+            )
+        else:
+            self._endpoint_state.delete_endpoint(deployment_name)
+
+        return updating
+
+    def apply_deployment_args(self, deployment_params: List[Dict]) -> None:
```

Comment:
This function will not work if the application under the deleting state. i think we can failed it loudly.